core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/AvroUpsertMonitor.scala

```diff
@@ -0,0 +1,234 @@
+        case _ => acknowledgeMessage(message.ackId).map(_ => None) pipeTo self //some other file (i.e. success/failure log, metadata.json)
+      }
+
+    case None =>
+      // there was no message so wait and try again
+      val nextTime = org.broadinstitute.dsde.workbench.util.addJitter(pollInterval, pollIntervalJitter)
+      system.scheduler.scheduleOnce(nextTime, self, StartMonitorPass)
+
+    case ImportComplete => self ! None
+
+    case Status.Failure(t) => throw t
+
+    case ReceiveTimeout => throw new WorkbenchException("AvroUpsertMonitorActor has received no messages for too long")
+
+    case x =>
+      logger.info(s"unhandled $x")
+      self ! None
+  }
+
+  private def initAvroUpsert(jobId: String, ackId: String): Future[Unit] = {
+    //start this Future asap because it can be long-running
+    val readUpsertObjectFuture = readUpsertObject(s"$jobId/upsert.json")
+
+    for {
+      avroMetadataJson <- readMetadataObject(s"$jobId/metadata.json")
+      //ack the response after we load the json into memory. pro: don't have to worry about ack timeouts for long operations, con: if someone restarts rawls here the uspert is lost
+      petSAJson <- samDAO.getPetServiceAccountKeyForUser(avroMetadataJson.namespace, RawlsUserEmail(avroMetadataJson.userEmail))
+      petUserInfo <- googleServicesDAO.getUserInfoUsingJson(petSAJson)
+      avroUpsertJson <- readUpsertObjectFuture
+      ackResponse <- acknowledgeMessage(ackId)
+      upsertResults <-
+        avroUpsertJson.grouped(batchSize).toList.traverse { upsertBatch =>
+          IO.fromFuture(IO(workspaceService.apply(petUserInfo).batchUpdateEntities(WorkspaceName(avroMetadataJson.namespace, avroMetadataJson.name), upsertBatch, true)))
+        }.unsafeToFuture
+    } yield {
+      logger.info(s"completed Avro upsert job ${avroMetadataJson.jobId} for user: ${avroMetadataJson.userEmail} with size ${avroUpsertJson.size} entities")
+      ()
+    }
+  }
+
+  private def acknowledgeMessage(ackId: String) =
+    pubSubDao.acknowledgeMessagesById(pubSubSubscriptionName, Seq(ackId))
+
+  private def parseMessage(message: PubSubMessage) = {
+    message.contents.parseJson.asJsObject.getFields("name").head.compactPrint match {
+      case objectIdPattern(jobId, file) => (jobId, file)
+      case m => throw new Exception(s"unable to parse message $m")
+    }
+  }
+
+  private def readUpsertObject(path: String): Future[Seq[EntityUpdateDefinition]] = {
+    readObject[Seq[EntityUpdateDefinition]](path)
+  }
+
+  private def readMetadataObject(path: String): Future[AvroMetadataJson] = {
+    readObject[AvroMetadataJson](path)
+  }
+
+  private def readObject[T](path: String)(implicit reader: JsonReader[T]): Future[T] = {
+    googleStorage.getBlobBody(GcsBucketName(avroUpsertBucketName), GcsBlobName(path)).compile.to[Array].unsafeToFuture().map { byteArray =>
+      byteArray.map(_.toChar).mkString.parseJson.convertTo[T]
+    }
+  }
```

Comment:
Do you know if this call is retrieving the gzipped payload, or if it is being [transcoded](https://cloud.google.com/storage/docs/transcoding) and you're receiving the plaintext payload? Ideally we'd be sending gzip over the network to cut down on egress costs and download time.

Reply:
It's really good that you called this out because I realize now in my testing I was actually receiving the plaintext payload...because the files I was placing in the bucket were plaintext, which is neither of the real use-cases.

Will look into getting this to work with gzipped payloads